### ISSUE TITLE
Set custom registryAuthorizer for Go client

### ISSUE THREAD

**Issue by rynowak** [helm/helm#12584]
## Summary

We're encountering issues doing a programmatic pull from `ghcr.io`. Specifically, the user might have cached or stale credentials in their docker credential store. The default Helm client will send these credentials even if they are known to be expired.

Our chart is public, and we would be unblocked if we could force the Helm client to use anonymous access.

## Solutions

I found this issue: https://github.com/helm/helm/issues/12144 which seems similar. 

If we could override the authorizer using options [here](https://github.com/helm/helm/blob/4e447d87cd3222aa5e5ea2b38170e9d80b99c0b4/pkg/registry/client.go#L82C1-L82C1) then I think this would do what we wanted.

Are you open to a PR that adds an option for this?

I'm thinking it will look something like this:

```go
// ClientOptRegistryAuthorizer returns a function that sets the authorizer used to access to the registry.
func ClientOptRegistryAuthorizer(authorizer auth.Client) ClientOption {
	return func(client *Client) {
		client.authorizer = authorizer
	}
}
```

**Comment by gjenkins8** [comment]
Thanks, a PR for this would be welcome I think. https://github.com/helm/helm/issues/12144 never received any follow up, so aged out.

---

> Our chart is public, and we would be unblocked if we could force the Helm client to use anonymous access.

Keep in mind, a PR to allow supplying a customer authorizer would only help SDK users. For the Helm CLI, this logic is going to be necessarily fixed.

**Comment by rynowak** [comment]
> Keep in mind, a PR to allow supplying a customer authorizer would only help SDK users. For the Helm CLI, this logic is going to be necessarily fixed.

This is OK for my usecase :+1:

PR is here: https://github.com/helm/helm/pull/12588

I verified this locally with the following code in my repo:

```go
hc := helm.Configuration{}

authorizer := &registryauth.Client{}
client, err := registry.NewClient(registry.ClientOptRegistryAuthorizer(authorizer))
if err != nil {
  return nil, err
}

hc.RegistryClient = client
```

**Comment by blampe** [comment]
I have a similar use case and I'm excited to see https://github.com/helm/helm/pull/12588 is targeted for the [3.15 / May 8](https://github.com/helm/helm/milestone/139) milestone. I see it was approved last year but not merged -- is there anything still holding it back?

**Comment by EronWright** [comment]
@gjenkins8 as you mentioned [here](https://github.com/helm/helm/issues/12584#issuecomment-1820340567), there's still need for a fix to to avoid using stale credentials. Is there another issue tracking that, or?
